### ISSUE TITLE
[REQUEST] Hot Corners feature to show only current application windows?

### ISSUE THREAD

**Issue by GalacticSun** [linuxmint/cinnamon#10988]
Please forgive me if this isn't the correct place to request a feature for Cinnamon / LM, but I wasn't sure where else to post this.

I used to use Mac OS and loved their "Exposé" / "Mission Control" feature, which is just like Linux Mint's `Hot Corners->Show All Windows` feature.  I use this Hot Corners feature in Linux Mint to quickly switch to other application windows in a visual way with my mouse by selecting the window to switch to.

However, Apple also has a slightly modified version of this feature that allows the user to show the windows of only the current application you're using.  They have it available as a Hot Corner option called "Application Windows".

I used to use the "Application Windows" hot corner option all the time and would find it tremendously handy to have in Cinnamon / Linux Mint.  Mainly for my web browser where it would allow me to quickly and easily switch between all open browser windows, and not see every single other open window I have open.

Would it please be possible to get such an option for hot corners in Cinnamon / Linux Mint, by chance?

Thanks very much and keep up the great work.

_Related:_
https://www.digitaltrends.com/computing/how-to-use-mac-hot-corners/
https://osxdaily.com/2021/06/24/use-hot-corners-mac/

**Comment by GalacticSun** [comment]
Actually, if this is a relatively simple Javascript change somewhere in Cinnamon, can you please let me know how to make this change and how to have it appear as an option for Hot Corners, by chance?  I can take a shot at making the change myself.  Thanks.

**Comment by GalacticSun** [comment]
I guess that's a "no", then?  Do you guys even take feature requests like this?

**Comment by zeigerpuppy** [comment]
Would love this feature too

**Comment by zeigerpuppy** [comment]
also see:
- https://forums.linuxmint.com/viewtopic.php?t=379430
- https://github.com/linuxmint/Cinnamon/issues/2117

**Comment by GalacticSun** [comment]
I'm just curious if this ever has a chance at being implemented.  If not, how would I go about trying to make this change myself?  Could someone please point me to the right place, where this code is located, perhaps?  Would it be possible via third party code?  How about if i support the Cinnamon project financially by donating something to get this feature implemented?

**Comment by ewmailing** [comment]
I also really want this feature. (I would also appreciate hints on how to get started implementing this ourselves.) Apple really nailed the feature with Exposé. This is the best/closest thing I've ever experienced on Linux, but without the per-application feature, it is that 'so close, yet so far' feeling. I find that the lack of this feature completely kills my ability to use the rest of this feature productively. It is really frustrating because it is soooo close.
 

The workaround of separating every application into its own workspace doesn’t work for me. Here is a simple illustrative example of how I use this feature.

Imagine I have two separate tasks: 
1. I am developing some code.
2. I am writing some emails.

I put each of these tasks into their own workspace.

For writing code, I have a couple of different apps open.
- IDE, with multiple windows open for different source files, debugger, compile log, etc.
- Web browser with multiple windows open to related things, such as useful code examples, algorithms I need to apply, , source repo
- Terminal windows open to different directories to easily access different source code or application resources or dev tools.

For emails:
- Mail client with the main inbox, the current message I’m reading, and message I’m composing.
- Web browser open to specification document, team planning docs, spreadsheets, research, etc. that are connected to the email I'm working on.

Frequently when writing code or writing email, I will have that window on one-side of my screen, side-by-side with a useful web browser window on the other side. 

Then frequently, I need to either swap the browser window for a different one, that is in that same workspace.

Showing ALL application windows is too overwhelming. There are too many windows to see because the combined total is too large, especially when I know I only need to choose between just the web browser windows.

And the separate workspace workaround doesn’t work for me because I want the different apps in the same workspace so I can put them side-by-side. And I want workspaces to keep separate tasks isolated from each other.

**Comment by ewmailing** [comment]
I've finally found where the relevant code is and managed to cobble something together that works. Please note, I am not a Mint Cinnamon developer, and implemented this in the worst brute force way. I found the files in /usr/share/cinnamon/js/ui and directly modified the scripts via sudo and kept relaunching cinnamon on my computer after every change. But what I have seems to work well, and would like to start discussion about how to get this accepted into the official repository.

I've committed into a feature branch named 'scaleapp-feature' at:
https://github.com/ewmailing/cinnamon/tree/scaleapp-feature
and specific commit at:
https://github.com/ewmailing/cinnamon/commit/dc195586b86176f074e841759681d61b84ab346a

This implements a variation based on the 'scale' feature that shows only the current  application windows instead of all the application windows. The new variation is called 'scaleapp'. It shares the 'scale' code and simply filters the list of windows down to the current application if this mode is requested. 

The changes are primarily in workspace.js, but also touches hotCorner.js and overview.js.

The changes in cs_hotcorner.py make the feature accessible in the UI. The name used is 'Show application windows', which is derived from the name 'Application Windows' from Mac, but follows the existing naming paradigm in the existing Cinnamon UI. I found out KDE Plasma already supports this feature too. They use the label: 'Present Windows - Current Application', which doesn't fit the Cinnamon naming paradigm as well.

I am using Mint 22.2 (the current official stable), so my script files are tied to that release, and not the main head. For specifics:
````
$ inxi -Sxxx
System:
  Host: LinuxRyzen Kernel: 6.17.0-23-generic arch: x86_64 bits: 64
    compiler: gcc v: 13.3.0 clocksource: tsc
  Desktop: Cinnamon v: 6.4.8 tk: GTK v: 3.24.41 wm: Muffin v: 6.4.1 vt: 7
    dm: LightDM v: 1.30.0 Distro: Linux Mint 22.2 Zara base: Ubuntu 24.04 noble
````

Being aware of this, I looked through the tags and checked out 6.4.8 and committed my modifications from that reference point. This should allow git to be able to merge forward when the time comes for any parts that aren't new code or direct conflicts. However, I noticed that there has already been a refactor of the Javascript API (most recent commits by @mtwebster ) so my new code will probably need to undergo similar changes. I repeat, I am not a Mint Cinnamon developer, and I was just hacking directly on the files installed on my system. I don't have an environment to actually develop for Mint Cinnamon, so it may make more sense for an actual maintainer to take the reins on this.

**Comment by GalacticSun** [comment]
Wow, @ewmailing, that is amazing, and thank you!

However, I wasn't able to get it to work.  I tried making those changes manually to those files (cs_hotcorner.py, hotCorner.js, overview.js, and workspace.js), and restarting Cinnamon, but it made it so the System Settings application won't open.  As a result, I can't set the new 'Show application windows' option in the System Settings Hotcorner UI.

I'm using Mint 22.3, so maybe there's a tweak I need to make somewhere?  Or something else I need to do to get it to work?
```
System:
  Host: computer Kernel: 6.8.0-117-generic arch: x86_64 bits: 64 compiler: gcc
    v: 13.3.0 clocksource: tsc
  Desktop: Cinnamon v: 6.6.7 tk: GTK v: 3.24.41 wm: Muffin v: 6.6.3 dm:
    1: LightDM v: 1.30.0 2: SDDM note: stopped Distro: Linux Mint 22.3 Zena
    base: Ubuntu 24.04 noble
```

**Comment by ewmailing** [comment]
@GalacticSun I'm pretty new to Mint and didn't realize point releases aren't part of the regular update process, and I didn't realize Mint 22.3 was already out. So if 22.3 is version 6.6.7, there have been some changes since 6.4.8. On quick visual inspection, they don't seem to be major changes, but some things were renamed which is probably sufficient to break your scripts if you blindly copy over your 22.3 scripts with my 22.2 scripts.

But I knew I was behind the master head, so I anticipated my changes would ultimately need to be merged forward, which is why I made sure to base my changes on the older 6.4.8, so they could easily be merged forward. So I merged forward my changes to 6.6.7. I've created a new topic branch named scaleapp-6.6.7. The merge was pretty clean. (There was some top level repo stuff for changelog and meson notes which I never touched or committed, but merge indicated a conflict which I didn't understand. But the correct files in this case are to fully take the 6.6.7 version, which I did.)

You can find my new 6.6.7 topic branch at:
https://github.com/ewmailing/cinnamon/tree/scaleapp-6.6.7
and the merge/commit:
https://github.com/ewmailing/cinnamon/commit/3c6ac836501c5c0c51c3a3d7f69e17528dd19d6f

Please let me know if this 6.6.7 based version works for you. I haven't tested this version because I need to backup my system first before I attempt to do an update to 22.3.

**Comment by ewmailing** [comment]
@GalacticSun Also, I forgot to say, try copying over all the files except cs_hotcorner.py first. Make sure the old scale mode still works before trying to change the HotCorners UI with the cs_hotcorner.py. So if there is a problem with the other scripts, then we can rule out this one.

And if you want to do some print debugging, you can sprinkle these over the code to see where things are breaking:
```
global.log("This is a debug message");
```
Alt-F2 will bring up a command launcher. Type:
```
lg
```
which will bring up Looking Glass app which shows the log messages.

The log messages also get written to ~/.xsession-errors if you break Cinnamon and have to fallback to a terminal. (All this is what I learned the hard way trying to develop this feature.)

**Comment by GalacticSun** [comment]
Okay, that is awesome.  I got it working!  Very nice work, @ewmailing!

The System Settings application not starting was actually due to me having made copies of the files and left them in their directories, I think, and Cinnamon apparently doesn't like that.  It works now, though.  Anyway, I removed those and added the 6.6.7 changes you made, and it works perfectly!

THANK YOU!  I've literally waited years for this and have tried a few times myself to get it working, but just couldn't quite figure it all out.  Very much appreciated.  I hope the Cinnamon devs merge in your work, because this is, in my opinion, a very nice addition!

**Comment by ewmailing** [comment]
@GalacticSun Terrific! Maybe you can try to grab the attention of the Cinnamon team. This bug might be so old, that they are ignoring it because it aged out in their filters. It has no labels or types, whereas some other bugs will at least be marked as Feature or Pull Request or something.

**Comment by GalacticSun** [comment]
Good point.  What do you think is a good way to go about getting their attention?  I can create another issue and reference this one, perhaps?

**Comment by ewmailing** [comment]
@GalacticSun I'm pretty new here, so I don't know how things usually go. But in the absence of any other ideas, I think that is a reasonable approach. Maybe you can describe the new one as a pull request since we now have code in this one.

**Comment by GalacticSun** [comment]
Ok, made a new Issue here:
https://github.com/linuxmint/cinnamon/issues/13783

Let me know if you think I should edit or add anything (I'm no Github expert!) or if I got anything wrong.
And feel free to add a comment or whatever to help boost the visibility, I guess.

Thanks again!

**Comment by ewmailing** [comment]
The merge conflict I got in branch scaleapp-6.6.7 bothered me, so I went back and redid it. The new merge is cleaner. The git tooling lied to me about using 'mine' vs. 'theirs'. Last time I said use theirs. but it appears I was supposed to use mine. The tooling perspective is reversed.

Anyway, I have a new topic branch called scaleapp-6.6.7b which has a cleaner merge and doesn't touch the changelog or meson files. Use that one for future merging and ignore the old scaleapp-6.6.7 branch.

https://github.com/ewmailing/cinnamon/tree/scaleapp-6.6.7b
https://github.com/ewmailing/cinnamon/commit/f0e61c5486e551095d500d17c756ce8faf2c1a92

**Comment by GalacticSun** [comment]
Ok, sounds good.
Do you mind updating that other [Issue](https://github.com/linuxmint/cinnamon/issues/13783) with that information, by chance?  I'm guessing the devs need to know that if they are going to merge the correct branch in?

**Comment by JosephMcc** [comment]
I closed the other issue because it's just spamming. We don't need more than one. Feature requests are no longer even taken here. If you want a discussion on this, open a proper pull request.

**Comment by GalacticSun** [comment]
> If you want a discussion on this, open a proper pull request.

Aha, got it.

Thanks for merging that in!

**Comment by ewmailing** [comment]
>> Feature requests are no longer even taken here. If you want a discussion on this, open a proper pull request. 
> 
> Aha, got it.
> 
> Thanks for merging that in!

No, I don't think he merged it.
So where did feature requests move to? 
And where is the documentation for the procedure to 'open a proper pull request'?

**Comment by GalacticSun** [comment]
> No, I don't think he merged it.

Oh, really?  =/

> So where did feature requests move to?

Good question, I'd also like to know.  I know they don't accept them via the [Linux Mint forums](https://forums.linuxmint.com/), either.  At least, not that the devs actually read or officially consider, as far as I know.

> And where is the documentation for the procedure to 'open a proper pull request'?

Also a good question.
Perhaps this [Issue](https://github.com/linuxmint/cinnamon/issues/13779) is relevant to this?

Thanks for bringing up these questions since I'm really not that familiar with Github.

**Comment by ewmailing** [comment]
Will you help me do the right procedure for doing this?

Meanwhile, it occurred to me that I should implement an Applet for this new scaleapp feature for consistency and completeness. I only use the hot corners myself, but I noticed that there is a tray Applet for every one of these triggers. So I basically duplicated the scale applet and tweaked it for scaleapp. If you want to test, you can copy the new directory directly to your /usr/share/cinnamon/applets/ directory.

https://github.com/ewmailing/cinnamon/commit/950b61ac4f4c9b8da5b54d37bdcca7c7d306da78

**Comment by GalacticSun** [comment]
> Will you help me do the right procedure for doing this?

Sure, where do we start?  Open a Pull Request?  (No idea how to do that.)

> Meanwhile, it occurred to me that I should implement an Applet for this new scaleapp feature for consistency and completeness. I only use the hot corners myself, but I noticed that there is a tray Applet for every one of these triggers. So I basically duplicated the scale applet and tweaked it for scaleapp. If you want to test, you can copy the new directory directly to your /usr/share/cinnamon/applets/ directory.
> 
> [ewmailing@950b61a](https://github.com/ewmailing/cinnamon/commit/950b61ac4f4c9b8da5b54d37bdcca7c7d306da78)

I tested it and it works great on my system!  Nice work.  Good idea to implement the applet for consistency and completeness.

Hopefully the devs can include your work in the next release.

**Comment by ewmailing** [comment]
> > Will you help me do the right procedure for doing this?
> 
> Sure, where do we start? Open a Pull Request? (No idea how to do that.)

Find info on the correct location and procedure so I can follow the steps. My hunch is I may need to do the pull request since it is my repo. But the procedure they want is just as if not more important. Some mods have a no patience/no tolerance policy with a short temper and will ignore you or work against you if you set them off. 

> 
> > Meanwhile, it occurred to me that I should implement an Applet for this new scaleapp feature for consistency and completeness. I only use the hot corners myself, but I noticed that there is a tray Applet for every one of these triggers. So I basically duplicated the scale applet and tweaked it for scaleapp. If you want to test, you can copy the new directory directly to your /usr/share/cinnamon/applets/ directory.
> > [ewmailing@950b61a](https://github.com/ewmailing/cinnamon/commit/950b61ac4f4c9b8da5b54d37bdcca7c7d306da78)
> 
> I tested it and it works great on my system! Nice work. Good idea to implement the applet for consistency and completeness.
> 
> Hopefully the devs can include your work in the next release.

Thanks for testing. Glad to get confirmation it worked. (FYI, I updated to 22.3 the other day, so I'm now on the same version as you.)

**Comment by GalacticSun** [comment]
> Find info on the correct location and procedure so I can follow the steps. My hunch is I may need to do the pull request since it is my repo. But the procedure they want is just as if not more important. Some mods have a no patience/no tolerance policy with a short temper and will ignore you or work against you if you set them off.

Okay, well I did some research:

I think you've done the main things so far, forking the repo, creating a branch, making and committing the changes, and pushing your changes to your Github fork.  Correct?

Beyond that it seems like it's necessary to create an Issue to link to the PR (maybe we can use this one, per that dev's comment about spamming)?

And then create a PR.  To open a Pull Request:
- Go to your forked repository on GitHub.
- Click the Contribute or Compare & pull request button that appears near the top.
- Ensure the base repository is linuxmint/cinnamon and the base branch is master (or the current development branch).
- Title Format: Use a concise, descriptive title.
- Description: Clearly explain the problem, the solution, and how to reproduce/test the fix. Mention any related issue numbers.

And, it seems like they use Matrix chat and encourage people to interact with the devs there (`#cinnamon:matrix.org`), if you happen to use Matrix (I don't).  I don't think that's required, but it might get their (more immediate) attention?

There's also a thing called [Draft PRs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests), which might be useful, I have no idea, so I'll leave that up to you to decide.

And, of course:
- One PR per Fix/Feature: Do not bundle unrelated changes into a single PR. This makes review easier and prevents partial reverts.
- Do proper testing, etc.  If you run the latest Linux Mint (which we are) or LMDE release you’re guaranteed everything will work.

More resources:
https://linuxmint-developer-guide.readthedocs.io/en/latest/

Hopefully that's helpful!


> Thanks for testing.

You're welcome.  Thanks for doing all that work!

**Comment by ewmailing** [comment]
Thanks for the research. I will go through it again more thoroughly a little later. But there is probably one issue:

> Ensure the base repository is linuxmint/cinnamon and the base branch is master (or the current development branch).

As I wrote earlier, my changes are against the current stable release, not the master development branch. I glanced at the master and they did some kind of renaming due to changes in their Javascript/GObject bindings a month ago. I'm concerned my new code will need to be manually migrated to the new API names. And as I stated earlier, I am just hacking directly on my system. I am not setup for proper Cinnamon head development. 

Is this something you might be able to do (or document for me)? While I can take a blind stab at migrating the code to the head, I don't have a testing environment. Would you be able to do the testing for me? They might have documentation on how to set up a testing environment on a stable Mint distribution...maybe you can find it and try to set it up? (If it's easy, you can show me how to do it.) Or as a fallback, maybe there is a way to get a nightly build of head that can be run on a spare computer or inside a VM?

**Comment by GalacticSun** [comment]
I don't think we need a specific dev testing environment, though.  As long as we're on the current release, I think that is sufficient.

And, it seems like the Applet addition might be better as a separate PR (and Issue).

Also, not sure if you've tested with Wayland or not, but I think you can log in using Wayland when you choose your session (when logging into LM).  There's a picker icon, I believe.

(Previous AI answer info deleted)

**Comment by ewmailing** [comment]
I think that AI answer is mostly BS and should be disregarded.

Ultimately, the problem is that master is working on the next Mint release, e.g. 22.4 or 23.0, and has decided breaking changes in these scripts are okay. So if I need to provide them a pull request based on master, then I need to make changes that will no longer work on our 22.3 machines.

So we need migrate my changes and test against their current master version of Mint Cinnamon. These cannot be tested and run as we've been doing on our 22.3 files by writing over /usr/share/cinnamon/, because the new system is incompatible. So we need to set up a proper environment.

I saw this link in your earlier write up. I think this is what we want. Can you try this?
https://linuxmint-developer-guide.readthedocs.io/en/latest/setup.html
I think it starts there but needs to keep going through:
https://linuxmint-developer-guide.readthedocs.io/en/latest/building.html
and
https://linuxmint-developer-guide.readthedocs.io/en/latest/romeo.html

I still don't have an intuition of how this actually produces a testing environment and how it doesn't break (or maybe it does) your regular stable Mint environment.

**Comment by GalacticSun** [comment]
Agreed.  And, to be perfectly honest, I'm not all that comfortable doing that on my system either because I know how Linux can be and things can get really messed up if you're not sure what you're doing with lower-level stuff like this.

That documentation is not that great in the sense that it just says to "install the devtools" and then `git clone https://github.com/linuxmint/mintsystem.git`, but, yes, as you said, does that mess up your current installation?  How does it affect the actual window rendering system?  ie. can you change files in there and suddenly X11 and Cinnamon starts using them?  I don't think so.  So, to me, that's confusing.

The distinction is also unclear between what appears to be "building applications" and editing JS files, etc, on the filesystem (https://linuxmint-developer-guide.readthedocs.io/en/latest/building.html, "Romeo", etc).  I get it if devs want to actually build (ie. compile) software, but that is different than editing JS files for the window rendering system in Cinnamon.

The only other way I can think of to test this would be to actually install 22.4 fresh on another computer with a display - because how else can the window rendering system pick up the changes to those JS files and display it all properly?

And, the date on those docs files are 2017 and also list applications that are [no longer in Mint at all](https://linuxmint-developer-guide.readthedocs.io/en/latest/mint-tools.html) (like MintNanny and MinUpload), so they clearly aren't prioritizing keep those docs up to date.

I still think this [Issue](https://github.com/linuxmint/cinnamon/issues/13779) is relevant (better docs / readme needed).  It seems to reveal that the devs aren't _really_ that interested in outside contributions.  Their current docs and communication tone seems to just be dismissive hand-wavey-ness expecting devs to already be kernel hackers and git experts.

I understand they're probably "busy", but there isn't really any excuse for such a lack of extensive documentation for a 20yr old OS (and one that seems to pull in a decent amount of money for a FOSS project).

**Comment by ewmailing** [comment]
Do you have a spare computer, or want to setup a virtual machine (e.g. VirtualBox or QEMU+Virtual Machine Manager)? I just tried installing VirtualBox, but I have some kind of incompatibility. I need to read up on QEMU+Virtual Machine Manager.

**Comment by GalacticSun** [comment]
Unfortunately, I don't have a spare machine.  I think you might have to turn off 3D rendering mode and have it only use software rendering.  Could that be the issue?

**Comment by ewmailing** [comment]
No, it has something to do with Virtualbox, AMD, and the kvm_amd Linux kernel module. I think it boils down to that Linux introduced its own KVM model, which I guess has specific vendor CPU implementations/extensions, and the kvm_amd doesn't work with Virtualbox because they did their own thing way before all of this, but didn't change their stuff to add support for it. The general guidance seems to be to disable or blacklist these kernel modules on my system, but I really don't want to mess with this stuff. The whole point of going to a VM was to NOT mess up my system. From what I'm reading, QEMU doesn't seem to have these issues.

**Comment by ewmailing** [comment]
I finally got QEMU+Virtual Machine Manager working and installed Mint 22.3 in it. Unfortunately, Cinnamon master fails to build with a compile error. (See bottom.)

I also tried rolling back to some of the tags to see if I could get anything to build:
master.mint22  - FAIL
6.7.1-unstable - FAIL
6.7.0-unstable - FAIL
6.6.8 - PASS

But I think 6.6.8 predates all the recent changes, so there is no point to using this one.

Even if I got it built, I'm not sure how to use it. But since I can't even get it built, it feels like hopelessness. So frustrating and this has been a massive waste of time.


As a final ditch effort, since I can make snapshots and clones with QEMU, I tried following the Romeo instructions to pull unstable packages. But after I enabled Romeo, I wasn't offered any new packages. Maybe Mint 22.3 is so new, they don't have anything yet they want to publish for Romeo?


Anyway, this is a dead end. I need to rest and regroup, but I think maybe I need to either make a pull request on what I already have, or take a blind shot at migrating it to master and make a pull request on that (knowing it will be untested).


Maybe one thing you can do to help is make a video recording of the new feature in action, to help explain it and show it off to why it is needed/helps, contrasted to the all-windows-variant. If you publish the video to YouTube, I can link it in the PR. Maybe the video could also include footage of Mac's in action and also KDE Plasma, showing that only Mint made the oversight to not implement the app-variant of this feature.




```
FAILED: src/libcinnamon.so.p/cinnamon-global.c.o 
ccache cc -Isrc/libcinnamon.so.p -Isrc -I../src -I. -I.. -Isrc/st -I../src/st -I../src/tray -I../src/st/croco -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/atk-1.0 -I/usr/include/cjs-1.0 -I/usr/include/gobject-introspection-1.0 -I/usr/include/cairo -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/muffin/clutter -I/usr/include/gio-unix-2.0 -I/usr/include/json-glib-1.0 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/fribidi -I/usr/include/muffin/cogl -I/usr/include/graphene-1.0 -I/usr/lib/x86_64-linux-gnu/graphene-1.0/include -I/usr/include/cinnamon-menus-3.0 -I/usr/include/gcr-3 -I/usr/include/p11-kit-1 -I/usr/include/gck-1 -I/usr/include/gtk-3.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/x86_64-linux-gnu -I/usr/include/webp -I/usr/include/muffin -I/usr/include/cinnamon-desktop -I/usr/include/polkit-1 -I/usr/include/xapp -I/usr/include/libxml2 -I/usr/include/libnm -I/usr/include/libsecret-1 -I/usr/include/gstreamer-1.0 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch '-DDATADIR="/usr/share"' '-DLIBDIR="/usr/lib/x86_64-linux-gnu"' '-DBINDIR="/usr/bin"' '-DCINNAMON_DATADIR="/usr/share/cinnamon"' '-DJSDIR="/usr/share/cinnamon/js"' '-DMUFFIN_TYPELIB_DIR="/usr/lib/x86_64-linux-gnu/muffin"' -DPOLKIT_VERSION=124 -Wno-deprecated-declarations -Wno-deprecated -Wno-declaration-after-statement -g -O2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -ffile-prefix-map=/home/ewing/Source/GIT/cinnamon223=. -flto=auto -ffat-lto-objects -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fdebug-prefix-map=/home/ewing/Source/GIT/cinnamon223=/usr/src/cinnamon-6.7.1 -Wdate-time -D_FORTIFY_SOURCE=3 -fPIC -pthread -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -isystem /usr/include/mozjs-115 -MD -MQ src/libcinnamon.so.p/cinnamon-global.c.o -MF src/libcinnamon.so.p/cinnamon-global.c.o.d -o src/libcinnamon.so.p/cinnamon-global.c.o -c ../src/cinnamon-global.c
../src/cinnamon-global.c: In function ‘cinnamon_global_set_cursor’:
../src/cinnamon-global.c:632:18: error: ‘META_CURSOR_NOT_ALLOWED’ undeclared (first use in this function); did you mean ‘CINNAMON_CURSOR_NOT_ALLOWED’?
  632 |       ret_curs = META_CURSOR_NOT_ALLOWED;
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
      |                  CINNAMON_CURSOR_NOT_ALLOWED
../src/cinnamon-global.c:632:18: note: each undeclared identifier is reported only once for each function it appears in
../src/cinnamon-global.c:635:18: error: ‘META_CURSOR_MOVE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NONE’?
  635 |       ret_curs = META_CURSOR_MOVE;
      |                  ^~~~~~~~~~~~~~~~
      |                  META_CURSOR_NONE
../src/cinnamon-global.c:638:18: error: ‘META_CURSOR_COPY’ undeclared (first use in this function); did you mean ‘META_CURSOR_BUSY’?
  638 |       ret_curs = META_CURSOR_COPY;
      |                  ^~~~~~~~~~~~~~~~
      |                  META_CURSOR_BUSY
../src/cinnamon-global.c:641:18: error: ‘META_CURSOR_NO_DROP’ undeclared (first use in this function); did you mean ‘META_CURSOR_DND_COPY’?
  641 |       ret_curs = META_CURSOR_NO_DROP;
      |                  ^~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_DND_COPY
../src/cinnamon-global.c:644:18: error: ‘META_CURSOR_POINTER’ undeclared (first use in this function); did you mean ‘META_CURSOR_NONE’?
  644 |       ret_curs = META_CURSOR_POINTER;
      |                  ^~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_NONE
../src/cinnamon-global.c:647:18: error: ‘META_CURSOR_S_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_SW_RESIZE’?
  647 |       ret_curs = META_CURSOR_S_RESIZE;
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_SW_RESIZE
../src/cinnamon-global.c:650:18: error: ‘META_CURSOR_N_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NW_RESIZE’?
  650 |       ret_curs = META_CURSOR_N_RESIZE;
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_NW_RESIZE
../src/cinnamon-global.c:653:18: error: ‘META_CURSOR_W_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NW_RESIZE’?
  653 |       ret_curs = META_CURSOR_W_RESIZE;
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_NW_RESIZE
../src/cinnamon-global.c:656:18: error: ‘META_CURSOR_E_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NE_RESIZE’?
  656 |       ret_curs = META_CURSOR_E_RESIZE;
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_NE_RESIZE
../src/cinnamon-global.c:674:18: error: ‘META_CURSOR_TEXT’ undeclared (first use in this function); did you mean ‘META_CURSOR_LAST’?
  674 |       ret_curs = META_CURSOR_TEXT;
      |                  ^~~~~~~~~~~~~~~~
      |                  META_CURSOR_LAST
../src/cinnamon-global.c:677:18: error: ‘META_CURSOR_GRABBING’ undeclared (first use in this function); did you mean ‘META_CURSOR_TRACKER’?
  677 |       ret_curs = META_CURSOR_GRABBING;
      |                  ^~~~~~~~~~~~~~~~~~~~
      |                  META_CURSOR_TRACKER
../src/cinnamon-global.c: In function ‘cinnamon_global_get_background_actors’:
../src/cinnamon-global.c:807:10: warning: implicit declaration of function ‘meta_get_background_actors_for_display’; did you mean ‘meta_get_x11_background_actor_for_display’? [-Wimplicit-function-declaration]
  807 |   return meta_get_background_actors_for_display (global->meta_display);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |          meta_get_x11_background_actor_for_display
../src/cinnamon-global.c:807:10: warning: returning ‘int’ from a function with return type ‘GList *’ {aka ‘struct _GList *’} makes pointer from integer without a cast [-Wint-conversion]
  807 |   return meta_get_background_actors_for_display (global->meta_display);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/cinnamon-global.c: In function ‘entry_cursor_func’:
../src/cinnamon-global.c:897:40: error: ‘META_CURSOR_TEXT’ undeclared (first use in this function); did you mean ‘META_CURSOR_LAST’?
  897 |                            use_ibeam ? META_CURSOR_TEXT : META_CURSOR_DEFAULT);
      |                                        ^~~~~~~~~~~~~~~~
      |                                        META_CURSOR_LAST
../src/cinnamon-global.c: In function ‘cinnamon_global_get_pointer’:
../src/cinnamon-global.c:1523:45: warning: passing argument 2 of ‘meta_cursor_tracker_get_pointer’ from incompatible pointer type [-Wincompatible-pointer-types]
 1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
      |                                             ^~~~~~
      |                                             |
      |                                             graphene_point_t * {aka struct _graphene_point_t *}
In file included from ../src/cinnamon-global.c:9:
/usr/include/muffin/meta/meta-cursor-tracker.h:54:70: note: expected ‘int *’ but argument is of type ‘graphene_point_t *’ {aka ‘struct _graphene_point_t *’}
   54 |                                                 int                 *x,
      |                                                 ~~~~~~~~~~~~~~~~~~~~~^
../src/cinnamon-global.c:1523:53: warning: passing argument 3 of ‘meta_cursor_tracker_get_pointer’ from incompatible pointer type [-Wincompatible-pointer-types]
 1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
      |                                                     ^~~~~~~~~
      |                                                     |
      |                                                     ClutterModifierType *
/usr/include/muffin/meta/meta-cursor-tracker.h:55:70: note: expected ‘int *’ but argument is of type ‘ClutterModifierType *’
   55 |                                                 int                 *y,
      |                                                 ~~~~~~~~~~~~~~~~~~~~~^
../src/cinnamon-global.c:1523:3: error: too few arguments to function ‘meta_cursor_tracker_get_pointer’
 1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/muffin/meta/meta-cursor-tracker.h:53:16: note: declared here
   53 | void           meta_cursor_tracker_get_pointer (MetaCursorTracker   *tracker,
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/cinnamon-global.c: In function ‘cinnamon_global_alloc_leak’:
../src/cinnamon-global.c:1891:16: warning: unused variable ‘ptr’ [-Wunused-variable]
 1891 |         gchar *ptr = g_strdup_printf ("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
      |                ^~~
../src/cinnamon-global.c: In function ‘cinnamon_global_get_stage_xwindow’:
../src/cinnamon-global.c:1920:12: warning: implicit declaration of function ‘meta_get_stage_xwindow’; did you mean ‘meta_focus_stage_window’? [-Wimplicit-function-declaration]
 1920 |     return meta_get_stage_xwindow (global->meta_display);
      |            ^~~~~~~~~~~~~~~~~~~~~~
      |            meta_focus_stage_window

```

**Comment by leigh123linux** [comment]
> I finally got QEMU+Virtual Machine Manager working and installed Mint 22.3 in it. Unfortunately, Cinnamon master fails to build with a compile error. (See bottom.)
> 
> I also tried rolling back to some of the tags to see if I could get anything to build: master.mint22 - FAIL 6.7.1-unstable - FAIL 6.7.0-unstable - FAIL 6.6.8 - PASS
> 
> But I think 6.6.8 predates all the recent changes, so there is no point to using this one.
> 
> Even if I got it built, I'm not sure how to use it. But since I can't even get it built, it feels like hopelessness. So frustrating and this has been a massive waste of time.
> 
> As a final ditch effort, since I can make snapshots and clones with QEMU, I tried following the Romeo instructions to pull unstable packages. But after I enabled Romeo, I wasn't offered any new packages. Maybe Mint 22.3 is so new, they don't have anything yet they want to publish for Romeo?
> 
> Anyway, this is a dead end. I need to rest and regroup, but I think maybe I need to either make a pull request on what I already have, or take a blind shot at migrating it to master and make a pull request on that (knowing it will be untested).
> 
> Maybe one thing you can do to help is make a video recording of the new feature in action, to help explain it and show it off to why it is needed/helps, contrasted to the all-windows-variant. If you publish the video to YouTube, I can link it in the PR. Maybe the video could also include footage of Mac's in action and also KDE Plasma, showing that only Mint made the oversight to not implement the app-variant of this feature.
> 
> ```
> FAILED: src/libcinnamon.so.p/cinnamon-global.c.o 
> ccache cc -Isrc/libcinnamon.so.p -Isrc -I../src -I. -I.. -Isrc/st -I../src/st -I../src/tray -I../src/st/croco -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/atk-1.0 -I/usr/include/cjs-1.0 -I/usr/include/gobject-introspection-1.0 -I/usr/include/cairo -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/muffin/clutter -I/usr/include/gio-unix-2.0 -I/usr/include/json-glib-1.0 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/fribidi -I/usr/include/muffin/cogl -I/usr/include/graphene-1.0 -I/usr/lib/x86_64-linux-gnu/graphene-1.0/include -I/usr/include/cinnamon-menus-3.0 -I/usr/include/gcr-3 -I/usr/include/p11-kit-1 -I/usr/include/gck-1 -I/usr/include/gtk-3.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/x86_64-linux-gnu -I/usr/include/webp -I/usr/include/muffin -I/usr/include/cinnamon-desktop -I/usr/include/polkit-1 -I/usr/include/xapp -I/usr/include/libxml2 -I/usr/include/libnm -I/usr/include/libsecret-1 -I/usr/include/gstreamer-1.0 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch '-DDATADIR="/usr/share"' '-DLIBDIR="/usr/lib/x86_64-linux-gnu"' '-DBINDIR="/usr/bin"' '-DCINNAMON_DATADIR="/usr/share/cinnamon"' '-DJSDIR="/usr/share/cinnamon/js"' '-DMUFFIN_TYPELIB_DIR="/usr/lib/x86_64-linux-gnu/muffin"' -DPOLKIT_VERSION=124 -Wno-deprecated-declarations -Wno-deprecated -Wno-declaration-after-statement -g -O2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -ffile-prefix-map=/home/ewing/Source/GIT/cinnamon223=. -flto=auto -ffat-lto-objects -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fdebug-prefix-map=/home/ewing/Source/GIT/cinnamon223=/usr/src/cinnamon-6.7.1 -Wdate-time -D_FORTIFY_SOURCE=3 -fPIC -pthread -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -mfpmath=sse -msse -msse2 -isystem /usr/include/mozjs-115 -MD -MQ src/libcinnamon.so.p/cinnamon-global.c.o -MF src/libcinnamon.so.p/cinnamon-global.c.o.d -o src/libcinnamon.so.p/cinnamon-global.c.o -c ../src/cinnamon-global.c
> ../src/cinnamon-global.c: In function ‘cinnamon_global_set_cursor’:
> ../src/cinnamon-global.c:632:18: error: ‘META_CURSOR_NOT_ALLOWED’ undeclared (first use in this function); did you mean ‘CINNAMON_CURSOR_NOT_ALLOWED’?
>   632 |       ret_curs = META_CURSOR_NOT_ALLOWED;
>       |                  ^~~~~~~~~~~~~~~~~~~~~~~
>       |                  CINNAMON_CURSOR_NOT_ALLOWED
> ../src/cinnamon-global.c:632:18: note: each undeclared identifier is reported only once for each function it appears in
> ../src/cinnamon-global.c:635:18: error: ‘META_CURSOR_MOVE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NONE’?
>   635 |       ret_curs = META_CURSOR_MOVE;
>       |                  ^~~~~~~~~~~~~~~~
>       |                  META_CURSOR_NONE
> ../src/cinnamon-global.c:638:18: error: ‘META_CURSOR_COPY’ undeclared (first use in this function); did you mean ‘META_CURSOR_BUSY’?
>   638 |       ret_curs = META_CURSOR_COPY;
>       |                  ^~~~~~~~~~~~~~~~
>       |                  META_CURSOR_BUSY
> ../src/cinnamon-global.c:641:18: error: ‘META_CURSOR_NO_DROP’ undeclared (first use in this function); did you mean ‘META_CURSOR_DND_COPY’?
>   641 |       ret_curs = META_CURSOR_NO_DROP;
>       |                  ^~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_DND_COPY
> ../src/cinnamon-global.c:644:18: error: ‘META_CURSOR_POINTER’ undeclared (first use in this function); did you mean ‘META_CURSOR_NONE’?
>   644 |       ret_curs = META_CURSOR_POINTER;
>       |                  ^~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_NONE
> ../src/cinnamon-global.c:647:18: error: ‘META_CURSOR_S_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_SW_RESIZE’?
>   647 |       ret_curs = META_CURSOR_S_RESIZE;
>       |                  ^~~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_SW_RESIZE
> ../src/cinnamon-global.c:650:18: error: ‘META_CURSOR_N_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NW_RESIZE’?
>   650 |       ret_curs = META_CURSOR_N_RESIZE;
>       |                  ^~~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_NW_RESIZE
> ../src/cinnamon-global.c:653:18: error: ‘META_CURSOR_W_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NW_RESIZE’?
>   653 |       ret_curs = META_CURSOR_W_RESIZE;
>       |                  ^~~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_NW_RESIZE
> ../src/cinnamon-global.c:656:18: error: ‘META_CURSOR_E_RESIZE’ undeclared (first use in this function); did you mean ‘META_CURSOR_NE_RESIZE’?
>   656 |       ret_curs = META_CURSOR_E_RESIZE;
>       |                  ^~~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_NE_RESIZE
> ../src/cinnamon-global.c:674:18: error: ‘META_CURSOR_TEXT’ undeclared (first use in this function); did you mean ‘META_CURSOR_LAST’?
>   674 |       ret_curs = META_CURSOR_TEXT;
>       |                  ^~~~~~~~~~~~~~~~
>       |                  META_CURSOR_LAST
> ../src/cinnamon-global.c:677:18: error: ‘META_CURSOR_GRABBING’ undeclared (first use in this function); did you mean ‘META_CURSOR_TRACKER’?
>   677 |       ret_curs = META_CURSOR_GRABBING;
>       |                  ^~~~~~~~~~~~~~~~~~~~
>       |                  META_CURSOR_TRACKER
> ../src/cinnamon-global.c: In function ‘cinnamon_global_get_background_actors’:
> ../src/cinnamon-global.c:807:10: warning: implicit declaration of function ‘meta_get_background_actors_for_display’; did you mean ‘meta_get_x11_background_actor_for_display’? [-Wimplicit-function-declaration]
>   807 |   return meta_get_background_actors_for_display (global->meta_display);
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>       |          meta_get_x11_background_actor_for_display
> ../src/cinnamon-global.c:807:10: warning: returning ‘int’ from a function with return type ‘GList *’ {aka ‘struct _GList *’} makes pointer from integer without a cast [-Wint-conversion]
>   807 |   return meta_get_background_actors_for_display (global->meta_display);
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../src/cinnamon-global.c: In function ‘entry_cursor_func’:
> ../src/cinnamon-global.c:897:40: error: ‘META_CURSOR_TEXT’ undeclared (first use in this function); did you mean ‘META_CURSOR_LAST’?
>   897 |                            use_ibeam ? META_CURSOR_TEXT : META_CURSOR_DEFAULT);
>       |                                        ^~~~~~~~~~~~~~~~
>       |                                        META_CURSOR_LAST
> ../src/cinnamon-global.c: In function ‘cinnamon_global_get_pointer’:
> ../src/cinnamon-global.c:1523:45: warning: passing argument 2 of ‘meta_cursor_tracker_get_pointer’ from incompatible pointer type [-Wincompatible-pointer-types]
>  1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
>       |                                             ^~~~~~
>       |                                             |
>       |                                             graphene_point_t * {aka struct _graphene_point_t *}
> In file included from ../src/cinnamon-global.c:9:
> /usr/include/muffin/meta/meta-cursor-tracker.h:54:70: note: expected ‘int *’ but argument is of type ‘graphene_point_t *’ {aka ‘struct _graphene_point_t *’}
>    54 |                                                 int                 *x,
>       |                                                 ~~~~~~~~~~~~~~~~~~~~~^
> ../src/cinnamon-global.c:1523:53: warning: passing argument 3 of ‘meta_cursor_tracker_get_pointer’ from incompatible pointer type [-Wincompatible-pointer-types]
>  1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
>       |                                                     ^~~~~~~~~
>       |                                                     |
>       |                                                     ClutterModifierType *
> /usr/include/muffin/meta/meta-cursor-tracker.h:55:70: note: expected ‘int *’ but argument is of type ‘ClutterModifierType *’
>    55 |                                                 int                 *y,
>       |                                                 ~~~~~~~~~~~~~~~~~~~~~^
> ../src/cinnamon-global.c:1523:3: error: too few arguments to function ‘meta_cursor_tracker_get_pointer’
>  1523 |   meta_cursor_tracker_get_pointer (tracker, &point, &raw_mods);
>       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /usr/include/muffin/meta/meta-cursor-tracker.h:53:16: note: declared here
>    53 | void           meta_cursor_tracker_get_pointer (MetaCursorTracker   *tracker,
>       |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../src/cinnamon-global.c: In function ‘cinnamon_global_alloc_leak’:
> ../src/cinnamon-global.c:1891:16: warning: unused variable ‘ptr’ [-Wunused-variable]
>  1891 |         gchar *ptr = g_strdup_printf ("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
>       |                ^~~
> ../src/cinnamon-global.c: In function ‘cinnamon_global_get_stage_xwindow’:
> ../src/cinnamon-global.c:1920:12: warning: implicit declaration of function ‘meta_get_stage_xwindow’; did you mean ‘meta_focus_stage_window’? [-Wimplicit-function-declaration]
>  1920 |     return meta_get_stage_xwindow (global->meta_display);
>       |            ^~~~~~~~~~~~~~~~~~~~~~
>       |            meta_focus_stage_window
> ```

The 6.7.x and master code is fine, you failed to update the requirements to 6.7.x (muffin etc).
Also there is no need to recompile cinnamon, the .js files aren't compiled code.

**Comment by ewmailing** [comment]
@leigh123linux We have a working implementation of this feature. My first implementation was on Mint 22.2, based on 6.4.8. I then updated the implementation for Mint 22.3, based on 6.6.7.

We would like to submit this for official inclusion, but the docs imply it should be against master. But there seemed to be a refactor of related files to this sub-system in master not long ago. I don't have any expectation that they would work if I overwrote my system files in /usr/share/cinnamon on my 22.3 system. So I can try to update my implementation again for the latest changes in master, but I don't know how to run and test it since my 22.3 system does not necessarily reflect what is in master.

We have been looking for docs on how to produce a test environment and also how to submit a pull request that complies with the rules & standards you want for the Mint Cinnamon project.  We obviously haven't found this, per your statement:

> The 6.7.x and master code is fine, you failed to update the requirements to 6.7.x (muffin etc).

I was following this page:
https://linuxmint-developer-guide.readthedocs.io/en/latest/building.html
but git clone cinnamon instead of mintinstall. And then invoke mint-build, which required sudo permissions, and seemed to pull a whole bunch of stuff automatically. 
I'm not sure where or how to update my requirements to 6.7.x. 


But maybe more to the point, can you advise: 
- would it be sufficient to submit the 6.6.7 based version for the pull request?
- or if it must be against master, I could make the changes I think are necessary, but can somebody with the proper master environment be able to help test and approve the changes?

Thank you

**Comment by ewmailing** [comment]
Just an update. I got cinnamon built and installed the debs, but it crashes immediately on startup, showing me the dialog: Cinnamon just crashed -> You are currently running in fallback mode.

It asks if I want to restart Cinnamon, with an additional option checkbox to disable downloaded applets, desklets, & extensions. Restarting, regardless of whether I click the checkbox, always crashes again and brings up the dialog again.

To build, I found this community tutorial from 13 years ago on how to build cinnamon, which includes the muffin stuff.
https://community.linuxmint.com/tutorial/view/955

In my QEMU VM which is a clean Mint 22.3 install, I cloned a fresh version of muffin and cinnamon (contains none of my changes). I followed the guide and installed all the packages. I then built muffin and installed the packages it created.

I then built cinnamon. I got a compiler error:
../src/cinnamon-global.c:10:10: fatal error: meta/meta-background-actor.h: No such file or directory 10 | #include <meta/meta-background-actor.h>

Searching the web for this, I learned I needed to install the package libmutter-dev. My Mint had libmutter-14-dev, so I installed that. When I tried to build again, it still didn't find the headers. So I created a symlink in /usr/include from mutter-14 to mutter, and that allowed the build to finish all the way through.

I installed all the cinnamon .debs. I restarted my system. And then I get the Cinnamon just crashed dialog.


An alternative thing I tried was to restore my old QEMU clean snapshot of 22.3, and copy over the latest js/ui files into my /usr/share/cinnamon/js/ui. That broke things. I tried with a minimal number of files and then copying more to see if things would start working again. I started with with just workspace.js, and then more files I know I needed to touch, and then eventually all of ui. This was broken in every configuration, so I don't think this method is going to work.

**Comment by GalacticSun** [comment]
Wow.  There obviously needs to be better documentation for this from the devs.  It shouldn't be this difficult to contribute to LM.  You sure have done a lot of work trying to get this to work, kudos to you!  I don't think I would have gotten that far.  How do the devs make changes to these JS files?  Any other LM contributors out there willing to help us?

**Comment by ewmailing** [comment]
Another update. I figured out there are downloadable builds of the .deb files on Github.
https://github.com/linuxmint/cinnamon/releases
https://github.com/linuxmint/muffin/releases

I downloaded the packages.tar.gz from both projects under the master.mint22 sections.
I extracted all the packages from both and put all the .deb files in a combined directory.

I rolled back my QEMU snapshot to a more pristine Mint 22.3 state.

In the combined directory, I ran:
```
sudo dpkg -i *.deb
```

There was an error about cinnamon-screensaver:

```
dpkg: regarding cinnamon_6.7.1_amd64.deb containing cinnamon:
 cinnamon breaks cinnamon-screensaver (<< 6.7)
  cinnamon-screensaver (version 6.6.1+zena) is present and installed.

dpkg: error processing archive cinnamon_6.7.1_amd64.deb (--install):
 installing cinnamon would break cinnamon-screensaver, and
 deconfiguration is not permitted (--auto-deconfigure might help)
(Reading database ... 541756 files and directories currently installed.)
Preparing to unpack cinnamon-common_6.7.1_all.deb ...
Unpacking cinnamon-common (6.7.1) over (6.6.7+zena) ...
Preparing to unpack cinnamon-dbg_6.7.1_amd64.deb ...
Unpacking cinnamon-dbg (6.7.1) over (6.6.7+zena) ...
Selecting previously unselected package cinnamon-doc.
Preparing to unpack cinnamon-doc_6.7.1_all.deb ...
Unpacking cinnamon-doc (6.7.1) ...
Preparing to unpack gir1.2-meta-muffin-0.0_6.7.1_amd64.deb ...
Unpacking gir1.2-meta-muffin-0.0:amd64 (6.7.1) over (6.6.3+zena) ...
Preparing to unpack libmuffin0_6.7.1_amd64.deb ...
Unpacking libmuffin0:amd64 (6.7.1) over (6.6.3+zena) ...
Preparing to unpack libmuffin-dev_6.7.1_amd64.deb ...
Unpacking libmuffin-dev:amd64 (6.7.1) over (6.6.3+zena) ...
Preparing to unpack muffin_6.7.1_amd64.deb ...
Unpacking muffin (6.7.1) over (6.6.3+zena) ...
Preparing to unpack muffin-common_6.7.1_all.deb ...
Unpacking muffin-common (6.7.1) over (6.6.3+zena) ...
Preparing to unpack muffin-dbg_6.7.1_amd64.deb ...
Unpacking muffin-dbg (6.7.1) over (6.6.3+zena) ...
dpkg: dependency problems prevent configuration of cinnamon-dbg:
 cinnamon-dbg depends on cinnamon (= 6.7.1); however:
  Version of cinnamon on system is 6.6.7+zena.

dpkg: error processing package cinnamon-dbg (--install):
 dependency problems - leaving unconfigured
Setting up cinnamon-doc (6.7.1) ...
Setting up muffin-common (6.7.1) ...
Processing triggers for libglib2.0-0t64:amd64 (2.80.0-6ubuntu3.8) ...
Setting up libmuffin0:amd64 (6.7.1) ...
Setting up muffin (6.7.1) ...
Setting up muffin-dbg (6.7.1) ...
Setting up gir1.2-meta-muffin-0.0:amd64 (6.7.1) ...
Setting up libmuffin-dev:amd64 (6.7.1) ...
Setting up cinnamon-common (6.7.1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
Processing triggers for libc-bin (2.39-0ubuntu8.7) ...
Processing triggers for mailcap (3.70+nmu1ubuntu1) ...
Processing triggers for gnome-menus (3.36.0-1.1ubuntu3) ...
Processing triggers for mate-menus (1.26.1+mint1) ...
Processing triggers for desktop-file-utils (0.27-2build1) ...
Processing triggers for man-db (2.12.0-4build2) ...
Errors were encountered while processing:
 cinnamon_6.7.1_amd64.deb
 cinnamon-dbg
```

So I went to https://github.com/linuxmint/cinnamon-screensaver/releases/tag/master.mint22 and downloaded packages.tar.gz for that. I installed those debs, and then I was able to install all the above.

But it still crashes like before.

I tried relaunching Cinnamon through the command line via
```
cinnamon --replace --display=:0
```

I saw the following message:
```
(cinnamon:4566): GLib-GIO-ERROR **: 00:30:36.659: Settings schema 'org.cinnamon.desktop.a11y.mouse' does not contain a key named 'dwell-click-mode-lock'
Trace/breakpoint trap (core dumped)

```

Doing a search on this error, led me to this page:
https://github.com/linuxmint/cinnamon/issues/7565

It implied I need to install cinnamon-desktop packages. So I went to https://github.com/linuxmint/cinnamon-desktop/releases/tag/master.mint22, and downloaded and installed the debs from there.

Cinnamon stopped crashing. A quick look at /usr/share/cinnamon/js/ui/workspace.js appears to be the latest version that is in master (with all the new refactoring changes I need to deal with).

So I think I finally have a usable test environment.

I'm really tired, so I'll probably work on the merge-forward/refactor fixes tomorrow.

**Comment by GalacticSun** [comment]
Nice job!  Sorry I couldn't be more help, but you seem to have done just fine without it.  Hopefully that test env works ok.

**Comment by ewmailing** [comment]
I migrated the changeset to master and tested it. Worked on the first shot.

I plan to push everything to my public repo later today. Then I will start writing up a formal pull request.

If you happen to have been working on the video I described earlier, now is the time :)

**Comment by ewmailing** [comment]
For reference, here is my work for the merge into the master branch:
https://github.com/ewmailing/cinnamon/commit/d7caf78148610701968b599724fe614947b3b7fd
in my topic branch:
https://github.com/ewmailing/cinnamon/tree/scaleapp-all-mastermerge

I ran out of time today, so didn't get to writing up a pull request. Will try to get to it tomorrow.

**Comment by GalacticSun** [comment]
> If you happen to have been working on the video I described earlier, now is the time :)

Sorry, no, I haven't been working on a video.  Hopefully they will be receptive with just the description, and maybe even testing it out.

Again, great work on this, I hope they are able to merge in your work.

**Comment by GalacticSun** [comment]
Any word on getting this merged?

**Comment by GalacticSun** [comment]
Hi @ewmailing , I created another Issue to try to get a dev's attention:
https://github.com/linuxmint/cinnamon/issues/13888
Not sure what else we need to do, quite frankly.

**Comment by GalacticSun** [comment]
Also posted here:
https://github.com/orgs/linuxmint/discussions/1614
